### PR TITLE
add clustering example with 50k points

### PR DESCRIPTION
**Pull request by almarklein**
issues: #2590

This adds an example that uses sklearn to cluster points based on their distribution. There are 4 plots for 4 different datasets. The algorithm can be selected (but be aware that some algorithms may take a looong time with 50k points.

### CODE REVIEW

**Comment by almarklein** [comment]
cc @bryevdv let me know if this is more or less what you intended

**Comment by almarklein** [comment]
bokeh_clustering_example

File: examples/plotting/file/clustering.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+http://scikit-learn.org/stable/auto_examples/cluster/plot_cluster_comparison.html
+"""
+
+import time
+
+import numpy as np
+
+from sklearn import cluster, datasets
+from sklearn.neighbors import kneighbors_graph
+from sklearn.preprocessing import StandardScaler
```
Comment (bryevdv) [inline examples/plotting/file/clustering.py]: probably worth a `try...except ImportError` with a specific message saying this particular example requires `sklearn` (use `from __future__ import print_function`)

File: examples/plotting/file/clustering.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+# Create clustering algorithms
+dbscan = cluster.DBSCAN(eps=.2)
+birch = cluster.Birch(n_clusters=2)
+two_means = cluster.MiniBatchKMeans(n_clusters=2)
+spectral = cluster.SpectralClustering(n_clusters=2,
+                                        eigen_solver='arpack',
+                                        affinity="nearest_neighbors")
+affinity_propagation = cluster.AffinityPropagation(damping=.9,
+                                                    preference=-200)
```
Comment (bryevdv) [inline examples/plotting/file/clustering.py]: I don't always think this sort of thing is worthwhile, but I do think this is nicer for an example:
```
dbscan               = cluster.DBSCAN(eps=.2)
birch                = cluster.Birch(n_clusters=2)
two_means            = cluster.MiniBatchKMeans(n_clusters=2)
affinity_propagation = cluster.AffinityPropagation(damping=.9, preference=-200)
spectral             = cluster.SpectralClustering(n_clusters=2, 
                                                  eigen_solver='arpack', 
                                                  affinity="nearest_neighbors")
```

File: examples/plotting/file/clustering.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    else:
+        y_pred = algorithm.predict(X)
+    
+    # Plot
+    p = figure(webgl=True, title=name, plot_width=PLOT_SIZE, plot_height=PLOT_SIZE)
+    p.scatter(X[:, 0], X[:, 1], color=colors[y_pred].tolist(), alpha=0.1,)
+    plots.append(p)
+
+# Genearate and show the plot
+box = VBox(HBox(*plots[:2]), HBox(*plots[2:]))
```
Comment (bryevdv) [inline examples/plotting/file/clustering.py]: This is a little bit "clever", it it can easily be made more straightforward, it should be.

**Comment by bryevdv** [comment]
Tested on OS X with Chrome, works great. A few comments about style other wise LGTM.

**Comment by almarklein** [comment]
Comments addressed.

**Comment by bryevdv** [comment]
@almarklein I merged master, will merge PR to dev 0.10 when tests pass.

**Comment by damianavila** [comment]
I did not check the conflict, but when I tested the @almarklein PR I had a conflict in `tests/travis/test.yaml`

**Comment by bryevdv** [comment]
Yes I resolved that conflict.

**Comment by almarklein** [comment]
Ok, great!

**Comment by damianavila** [comment]
Well, there is still a conflict... let me check...

**Comment by damianavila** [comment]
OK, fixed... it needed merge-in dev_0.10, should be ready after travisCI run finish...

**Comment by damianavila** [comment]
And now the diff is reasonable :wink:

**Pull request merged**